### PR TITLE
Fixes bugs around timestamp initialization and comparison.

### PR DESCRIPTION
**Pull request by tgregg**
Timestamps with less than full precision have implicit defaults (1 for month and day, 0 for everything else).

When comparing equality, fields outside the stated precision are ignored.

### CODE REVIEW

File: ionc/ion_timestamp.c [hunk 1]
```diff
@@ -1056,6 +1056,16 @@ iERR _ion_timestamp_equals_helper(const ION_TIMESTAMP *ptime1, const ION_TIMESTA
...
         IONCHECK(_ion_timestamp_to_utc(ptime1, &ptime1_compare));
         IONCHECK(_ion_timestamp_to_utc(ptime2, &ptime2_compare));
+        if    ((ptime1_compare.year != ptime2_compare.year)
+               || (ptime1_compare.month != ptime2_compare.month)
+               || (ptime1_compare.day != ptime2_compare.day)
+               || (ptime1_compare.hours != ptime2_compare.hours)
+               || (ptime1_compare.minutes != ptime2_compare.minutes)
+               || (ptime1_compare.seconds != ptime2_compare.seconds)
+               || (ptime1_compare.tz_offset != ptime2_compare.tz_offset)
+                ) {
```
Comment (pbcornell) [inline ionc/ion_timestamp.c]: funky location of `) {` -- consider moving to previous line

File: ionc/ion_timestamp.c [hunk 2]
```diff
@@ -1074,25 +1084,34 @@ iERR _ion_timestamp_equals_helper(const ION_TIMESTAMP *ptime1, const ION_TIMESTA
...
         }
+        if (HAS_TZ_OFFSET(ptime1) && (ptime1->tz_offset != ptime2->tz_offset)) {
+            goto is_false;
+        }
+        if (precision1 >= ION_TS_YEAR) {
+            if (ptime1->year != ptime2->year) {
+                goto is_false;
+            }
+        }
+        else {
```
Comment (pbcornell) [inline ionc/ion_timestamp.c]: Consider moving this above the `precision1 >= ION_TS_YEAR` check above.  I.e.:
```
if (precision1 < ION_TS_YEAR) {
    FAILWITHMSG(...);
}
if (precision1 >= ION_TS_YEAR) {
   ...
```

This groups all the `precision1 >= ION_TS_...` checks together, and reads more easily.

File: ionc/ion_timestamp.c [hunk 3]
```diff
@@ -1074,25 +1084,34 @@ iERR _ion_timestamp_equals_helper(const ION_TIMESTAMP *ptime1, const ION_TIMESTA
...
+            FAILWITHMSG(IERR_INVALID_TIMESTAMP, "Found timestamp with less than year precision.");
+        }
+        if (precision1 >= ION_TS_MONTH && (ptime1->month != ptime2->month)) {
+            goto is_false;
+        }
+        if (precision1 >= ION_TS_DAY && (ptime1->day != ptime2->day)) {
+            goto is_false;
+        }
+        if (precision1 >= ION_TS_MIN && (ptime1->hours != ptime2->hours || ptime1->minutes != ptime2->minutes)) {
```
Comment (pbcornell) [inline ionc/ion_timestamp.c]: I'm surprised to find there's no `ION_TS_HOUR` precision.  Should there be?

File: ionc/ion_timestamp.c (hunk 1 above)
Comment (tgregg) [inline ionc/ion_timestamp.c]: Agreed

File: ionc/ion_timestamp.c (hunk 2 above)
Comment (tgregg) [inline ionc/ion_timestamp.c]: Sure

File: ionc/ion_timestamp.c (hunk 3 above)
Comment (tgregg) [inline ionc/ion_timestamp.c]: Nope. There is no such thing as hour precision -- if a timestamp has hour precision, it MUST have minute precision. Minute precision is really hour/minute precision.

**Pull request merged**